[New] `require-default-props`: add option `functions` (#3249)
jsx-eslint/eslint-plugin-react: docs/rules/require-default-props.md

```js
...
"react/require-default-props": [<enabled>, {
  "forbidDefaultForRequired": <boolean>,
  "ignoreFunctionalComponents": <boolean>,
  "forceDefaultPropArguments": <boolean>
}]
...
```

- `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
- `forbidDefaultForRequired`: optional boolean to forbid prop default for a required prop. Defaults to false.
- `ignoreFunctionalComponents`: optional boolean to ignore this rule for functional components. Defaults to false.

> - `forceDefaultPropArguments`: optional booolean to assign default to object instead of `defaultProps` for functional components. Defaults to false.

Reviewer: i think we should make sure that `forceDefaultPropArguments` can not be set to `true` when `ignoreFunctionalComponents` is `true`?

or maybe, we should deprecate `ignoreFunctionalComponents`, and come up with an alternate schema:
```
{
  "forbidDefaultForRequired": <boolean>,

  // for now, this wouldn't support any other choices, and would be optional, but later could be set to "ignore" as long as `functions` wasn't set to "ignore"
  "classes": "defaultProps",

  "functions": "ignore" | "defaultProps" | "default arguments"
}
```

Thoughts?

Author: > we should deprecate ignoreFunctionalComponents, and come up with an alternate schema

I like this, then should the schema be like following lines? 

```
{
  "forbidDefaultForRequired": <boolean>,
  "classes": "defaultProps",
  "functions": "ignore" | "defaultProps" | "defaultArguments"
}
```

Or should I keep the `forbidDefaultForRequired`?

Reviewer: Yep, exactly. And the new stuff would only be permitted if ignoreFunctionalComponents is omitted.

```js
...
"react/require-default-props": [<enabled>, {
  "forbidDefaultForRequired": <boolean>,
  "ignoreFunctionalComponents": <boolean>,
  "assignDefaultToObjectForFc": <boolean>
}]
...
```

- `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.

> - `forbidDefaultForRequired`: optional boolean to forbid prop default for
>   a required prop. Defaults to false.
> - `ignoreFunctionalComponents`: optional boolean to ignore this rule for
>   functional components. Defaults to false.
> - `assignDefaultToObjectForFc`: optional booolean to assign default to object
>   instead of `defaultProps` for functional components. Defaults to false.

Reviewer: ```suggestion
- `forbidDefaultForRequired`: optional boolean to forbid prop default for a required prop. Defaults to `false`.
- `ignoreFunctionalComponents`: optional boolean to ignore this rule for functional components. Defaults to `false`.
- `assignDefaultToObjectForFc`: optional boolean to assign default to object instead of `defaultProps` for functional components. Defaults to `false`.
```

```js
...
"react/require-default-props": [<enabled>, {
  "forbidDefaultForRequired": <boolean>,
  // For now, this wouldn't support any other choices, and would be optional,
  // but later could be set to "ignore" as long as `functions` wasn't set to "ignore"
  "classes": "defaultProps",
  "functions": "defaultProps" | "defaultArguments" | "ignore"
  // @deprecated Use `functions` option instead.
  "ignoreFunctionalComponents": <boolean>,
}]
...
```

- `enabled`: for enabling the rule. 0=off, 1=warn, 2=error. Defaults to 0.
- `forbidDefaultForRequired`: optional boolean to forbid prop default for a required prop. Defaults to false.

> - `classes`: For now, only works with "defaultProps".

Reviewer: not just for now, really? classes can only ever have `defaultProps`.

> - `functions`: optional string that this rule determines strategy for the function. "defaultArguments" is recommended. We will change it to deafult later. Defaults to "defaultProps".

Reviewer: defaultArguments is very much NOT recommended; defaultProps is a far superior mechanism.
```suggestion
- `functions`: optional string to determine which strategy a functional component uses for defaulting props. Defaults to "defaultProps".
```

Author: According to the original issue:
> defaultProps on function components is getting deprecated by React (see: facebook/react#16210)

It has been three-ish years since the original issue was reported. Is there any reason why the `defaultProps` is better now?

Reviewer: Three-ish years and no actual deprecation, yes.

The primary reason is the same: `.defaultProps` are introspectable from outside the component, and default arguments are not.

Author: Oh, I didn't think that far. Thank you.

};
```

### `functions`

- "defaultProps": Use `.defaultProps`.
- "defaultArguments": Use object default arguments
- "ignore": Ignore this rule for functional components.

> "defaultArguments" is recommended. We will change it to deafult later.

Reviewer: see above: it's not, and i don't plan to ever change the default. `.defaultProps` will remain the superior approach for the foreseeable future.

Examples of **incorrect** code for this rule:

```jsx
function Hello({ foo }) {
  return <div>{foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string
};
Hello.defaultProps = {
  foo: 'foo'
}
```

```jsx
function Hello({ foo = 'foo' }) {
  return <div>{foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string
};
Hello.defaultProps = {
  foo: 'foo'
}
```

```jsx
function Hello(props) {
  return <div>{props.foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string
};
```

Examples of **correct** code for this rule:

```jsx
function Hello({ foo = 'foo' }) {
  return <div>{foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string
};
```

```jsx
function Hello({ foo }) {
  return <div>{foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string.isRequired
};
```

```jsx
function Hello(props) {
  return <div>{props.foo}</div>;
}

Hello.propTypes = {
  foo: PropTypes.string.isRequired
};
```

### `ignoreFunctionalComponents`

When set to `true`, ignores this rule for all functional components.

> **Depreacted**, use `functions` instead.

Reviewer: ```suggestion
**Deprecated**, use `functions` instead.
```
